### ISSUE TITLE
Race condition registering event handlers for requests?

### ISSUE THREAD

**Issue by Pascal666** [w3c/IndexedDB#233]
I was unable to find anything in the spec that states what should happen if a request is ready to call its event handlers but they have not yet been registered.  I'm assuming it should then fire the event as soon as an appropriate handler is registered.

**Comment by Pascal666** [comment]
Thank you.  The pattern I normally see is something like:
```
var request = store.get(12345);
request.onsuccess = function() { <do something> };
```
Because the request is handled asynchronously, it is possible for the request to complete before the above event handler gets registered, yes?  I don't see how to register an event handler for a request before the request is created, so does this not create a race condition?  If the request completes before the event handler is registered then you are saying the event handler will never fire?

**Comment by brodybits** [comment]
I don't think so because according to [[1]](https://w3c.github.io/IndexedDB/#asynchronously-execute-a-request) each request goes into a queue [[2]](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task) for an event loop [[3]](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop) with a task queue [[4]](https://html.spec.whatwg.org/multipage/webappapis.html#task-queue).

There is a number of excellent resources that describe how JavaScript event loops work on the browser such as [[5]](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop), [[6]](https://blog.sessionstack.com/how-javascript-works-event-loop-and-the-rise-of-async-programming-5-ways-to-better-coding-with-2f077c4438b5), and [others (7)](https://www.google.com/search?q=web+event+loop). Node.js event loop is very similar, excellent resource at [[8]](https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/) along with [others (9)](https://www.google.com/search?ei=JPGyWtmaA4zy5gLUgY3wBw). Also looks like a really nice video at [[10]](https://www.youtube.com/watch?v=8aGhZQkoFbQ).

I hope this is enough information to clear up the questions raised, please speak up if you have any more questions.

[1] <https://w3c.github.io/IndexedDB/#asynchronously-execute-a-request>
[2] <https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task>
[3] <https://html.spec.whatwg.org/multipage/webappapis.html#event-loop>
[4] <https://html.spec.whatwg.org/multipage/webappapis.html#task-queue>
[5] <https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop>
[6] <https://blog.sessionstack.com/how-javascript-works-event-loop-and-the-rise-of-async-programming-5-ways-to-better-coding-with-2f077c4438b5>
[7] <https://www.google.com/search?q=web+event+loop>
[8] <https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/>
[9] <https://www.google.com/search?ei=JPGyWtmaA4zy5gLUgY3wBw>
[10] <https://www.youtube.com/watch?v=8aGhZQkoFbQ>
